catlearn/utils/metrics.py

```diff
@@ -0,0 +1,190 @@
+    Hence will default to default_rank above it
+    (if None all ranks are taken into account).
+    """
+    # compute ranks for each triplet
+    rank_lists = {
+        triplet[0]: list(cache.sort_relations(*triplet[1], n_items=max_rank))
+        for triplet in triplets
+    }
+
+    ranks = {}
+    for triplet, rank_list in tqdm(rank_lists.items()):
+        triplet_rank = get_index_in_list(
+            rank_list, triplet, default_index=default_rank,
+        )
+        ranks[triplet] = triplet_rank
+    return ranks
+
+
+def compute_kge_metrics(
+    cache: RelationCache[NodeType, ArrowType],
+    triplets: List[TripletMatchPair],
+    default_rank: int,
+    max_rank: Optional[int] = None,
+) -> Mapping[str, Any]:
+    """
+    Compute KGE metrics for 
+    - a list of tuples containing:
+        - an evaluation triplet of the form (src, tar, label) with the actual values
+        - a request triplet of the form (src, tar, label) where any of
+        the 3 values may be None, meaning we want to find a match for it
+    - default_rank: the default value of rank when the triplet is not found
+    - max_rank: only ranks up to max_rank are computed. Hence will default
+    to default_rank above it (if None all ranks are taken into account)
+    """
+    # compute ranks for each triplet
+    ranks = compute_eval_triplet_ranks(
+        cache, triplets, default_rank, max_rank=max_rank)
+    nb_triplets = len(triplets)
+
+    mean_rank = sum(ranks.values())/nb_triplets
+    mean_reciprocal_ranks = sum(
+        1./(1. + value) for value in ranks.values())/nb_triplets
+    max_found = max((value for value in ranks.values() if value < default_rank), default=default_rank)
+    hits_at_n = sum((
+        Tsor([float(i >= value) for i in range(max_found + 1)])
+        for value in ranks.values()
+    ))
+
+    return {
+        'mean_rank': mean_rank,
+        'mean_reciprocal_ranks': mean_reciprocal_ranks,
+        'hits_at_n': hits_at_n,
+    }
+
+
+def compute_average_missing_source_target_kge_metrics(
+    cache: RelationCache[NodeType, ArrowType],
+    triplets: List[Tuple[
+            Optional[NodeType], Optional[NodeType],
+            Optional[FrozenSet[ArrowType]]]],
+    default_rank: int,
+    max_rank: Optional[int] = None,
+) -> Mapping[str, Any]:
+    """
+    Log KGE metrics for given triplets:
+    remove source, then target and make prediction.
+    Average results of both
+    - default_rank: the default value of rank when the triplet is not found
+    - max_rank: only ranks up to max_rank are computed. Hence will default
+    to default_rank above it (if None all ranks are taken into account)
+    """
+    sourceless_triplets = (
+        (None, target, frozenset([label]))
+        for (source, target, label) in triplets
+    )
+    targetless_triplets = (
+        (source, None, frozenset([label]))
+        for (source, target, label) in triplets
+    )
+    full_triplets = [
```

Comment:
Better to use a generator here as well

Reply:
It is used twice, so it would be consumed the first time.
Hence 2 solutions:
- make it a list (or some other data structure); that's the current version
- make it a function with no args returning the iterator

Reply:
Right, I forgot about the consumed twice part. OK for me in the current implemntation then